File: graph_inference/solve_cascade.py
#!/usr/bin/env python
from __future__ import print_function
import csv
import numpy as np
import cvxpy as CVX
from collections import defaultdict

# This file has been changed from a script friendly format to an interactive
# session friendly version.

## You have to make sure to set these parameters to be the same as the
# parameters used to generate cascades.csv
input_file = open('cascades.csv', 'r')
time_period = 1.0
num_nodes = 50

# Cascades is a dictionary which will map cascade indexes (integers)
# to list of tuples of (time_of_infection, node_id_of_infected node)
cascades = defaultdict(lambda : [])

# Reading data
for row in csv.DictReader(input_file):
    # keys = 'cascade_id', 'dst', 'at'
    cascade_id = row['cascade_id']
    dst        = int(row['dst'])
    at         = float(row['at'])
    assert at < time_period, "Infection after observation period."
    cascades[cascade_id].append((at, dst))

# Start definition of the problem

# Sort according to time
for cascade_id in cascades.keys():
    cascades[cascade_id] = sorted(cascades[cascade_id])

# Possible edges: if we have never seen any infection travel down an edge, the
# best estimate for alpha for that edge is 0, i.e. it does not exist.
possible_edges = set()
for c in cascades.values():
    for i in range(len(c)):
       for j in range(i):
           possible_edges.add((c[j][1], c[i][1]))

# Formulating the problem for each row of influence matrix A
A = np.zeros((num_nodes, num_nodes), dtype=float)
probs = []
results = []


def logSurvival(t_i, t_j, alpha_ji):
    # TODO
    raise NotImplementedError('logSurvival not implemented.')

def hazard(t_i, t_j, alpha_ji):
    # TODO
    raise NotImplementedError('hazard not implemented.')

# These problems can be solved in parallel
for target_node in range(num_nodes):

    # This is one column of the alpha matrix
    Ai = CVX.Variable(num_nodes, name='A[:, {}]'.format(target_node))

    # Which edges are constrained to be zero
    constraints = []
    for j in range(num_nodes):
        if (j, target_node) not in possible_edges:
            constraints.append(Ai[j] == 0)
        else:
            constraints.append(Ai[j] >= 0)

    # Constructing the log-likelihood for the target_node
    expr = 0
    for c_idx, c in cascades.items():
        infection_time_arr = [x[0] for x in c if x[1] == target_node]

        assert len(infection_time_arr) <= 1

        if len(infection_time_arr) == 0:
            # Node 'i' was not infected in this cascade
            for j in range(len(c)):
                alpha_ji = Ai[c[j][1]]
                t_j = c[j][0]
                T = time_period
                expr += logSurvival(T, t_j, alpha_ji)
        else:
            # Node 'i' was infected in this cascade
            infection_time = infection_time_arr[0]
            t_i = infection_time

            if c[0][0] != infection_time:
                # Do this only if the target_node wasn't the first node
                # infected in the cascade.
                # If it was the first node (else branch), then we cannot deduce
                # anything about the incoming edges.
                log_sum = 0
                for j in range(len(c)):
                    # TODO => Complete the Log-Likelihood if
                    # the node was infected.
                    pass

                expr += CVX.log(log_sum)

    prob = CVX.Problem(CVX.Maximize(expr), constraints)
    res = prob.solve(verbose=True)
    probs.append(prob)
    results.append(res)
    if prob.status in [CVX.OPTIMAL, CVX.OPTIMAL_INACCURATE]:
        A[:, target_node] = np.asarray(Ai.value).squeeze()
    else:
        A[:, target_node] = -1

File: graph_inference/solve_cascade.m
C = csvread('cascades.csv', 1); % Read the CSV, skipping the header line
                                % C(:, 1) = cascade_id; C(:, 2) = dest_id; C(:, 3) = infection time
num_nodes = 50;
time_period = 1.0;

% Assuming that the cascades are sorted by time.
% Assuming that the nodes are numbered sequentially.

cascade_ids = unique(C(:, 1));

% Not all the edges are possible (if two nodes do not appear toguether in
% any cascade, the MLE solution for this alpha is zero)
possible_edges_arr = [];
for c_idx = 1:size(cascade_ids, 1)
    cascade = C(C(:, 1) == c_idx, :);
    for ii = 1:size(cascade, 1)
        for j = 1:(ii - 1)
            possible_edges_arr(end + 1, :) = [cascade(j, 2)+1, cascade(ii, 2)+1];
        end
    end
end

possible_edges = unique(possible_edges_arr, 'rows');

A = zeros(num_nodes, num_nodes); %influence parameters

for target_node = 1:num_nodes % Distributed
    cvx_begin % Optimization solver
        variable Ai(num_nodes);
        expression expr;

        for c_idx = 0:size(cascade_ids, 1)-1
            cascade = C(C(:, 1) == c_idx, :);
            infection_time = cascade(cascade(:, 2)+1 == target_node, 3);

            if size(infection_time, 1) == 0 % The node wasn't infected
                for j = 1:size(cascade, 1)
                    expr = expr + % TODO: Survival (log(S(T | t_j, alpha_ji)))
                end
            else % The node was infected
                if cascade(1, 3) ~= infection_time
                    % Do this only if this node wasn't the first node infected.
                    % If this was the first node infected, then we cannot
                    % deduce anything about the incoming edges.
                    expression log_sum;

                    for j = 1:size(cascade, 1)
                        % TODO ==> Complete the  Log-Likelihood
                    end

                    expr = expr + log(log_sum);
                end
            end
        end

        maximize expr
        subject to
            for j = 1:num_nodes
                if ismember([j, target_node], possible_edges)
                    Ai(j) >= 0
                else
                    Ai(j) == 0
                end
            end
    cvx_end

    A(:, target_node) = Ai;
end

A_soln = dlmread('solution.csv'); % This is the true A matrix.

[F1, precision, recall] = calc_score(A, A_soln, 1e-6)
